subprojects/core/src/main/java/org/gradle/internal/build/DeferredRootBuildGradle.java

```diff
@@ -0,0 +1,44 @@
+/*
+ * Copyright 2024 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.build;
+
+import org.gradle.api.NonNullApi;
+import org.gradle.api.internal.GradleInternal;
+import org.gradle.internal.service.scopes.Scope;
+import org.gradle.internal.service.scopes.ServiceScope;
+
+/**
+ * Access to the root build's {@link GradleInternal} instance
+ * that becomes available as soon as the root build state is created.
+ */
+@NonNullApi
+@ServiceScope(Scope.BuildTree.class)
+public interface DeferredRootBuildGradle {
```

Comment:
❓ If we consider the late binding a temporary measure, I'd prefer us not exposing the class outside CC. We can have a CC-specific implementation of `BuildLifecycleControllerFactory` that does necessary decorations to have that.

Reply:
Right. It could be decoration based or template-method based but it does make sense to keep CC-only for now.

Reply:
I moved the service class into `configuration-cache`